동기화 메모장 [STEP2] Jiseong, Toni, 애플사이다

just1103 opened the pull request:
안녕하세요 콘! @protocorn93
동기화 메모장 STEP2 완료하여 PR 제출드립니다!
STEP1 때 코드 구조 관련 코멘트를 주셔서 많은 도움이 되었어요. 이번 리뷰도 잘 부탁드립니다. 😁

## STEP2 구현내용 및 고민한 점
- 싱글톤 `CoreDataManager`을 통해 CoreData CRUD 기능을 구현했습니다.
- `메모 우상단의 더보기 버튼`과 `Cell Swipe 버튼`을 통해 메모 공유/삭제 기능을 구현했습니다.
- TextView의 메모 내용 중 첫 줄은 Title이 되고, 줄바꿈 이후부터 Body가 되도록 구현했습니다.
- STEP2 요구사항에는 없지만 앱 테스트를 원활히 진행하고자 TableView 상단의 `+버튼`을 통해 새 메모를 추가하도록 임시 기능을 구현했습니다.

### 1. 범용성 있는 CoreDataManager
- 메모 데이터를 담고 있는 `memos 배열`을 CoreData와 TableViewDataSource 중에 어디에 저장해야 할지에 대해 고민했습니다.
- CoreDataManager를 범용성 있는 Utility 기능으로 구현하고 싶었기 때문에 CoreData가 memos를 가지지 않도록 하는 게 적절하다고 판단했습니다. 따라서 TableViewDataSource가 가지도록 했습니다. 또한 CoreDataManager의 fetch/delete 메서드 등도 확장성을 고려하여 제네릭 타입으로 생성했습니다.

### 2. MasterTableViewController 및 DetailViewController의 관계
- DetailView에서 받은 사용자 입력값을 어떻게 TableView List에 반영할지에 대해 고민했습니다. SplitViewController에 담겨있는 Master 및 Detail View의 의존성을 낮추기 위해 서로 모르도록 했습니다. 따라서 사용자가 DetailView에서 텍스트를 입력하거나, 메모를 삭제할 때 NofiticationCenter를 사용하여 MasterView에게 알려주도록 했습니다.
- 그런데 기능을 추가하다보니 프로젝트 특성상 두 View의 연관성이 높을 수 밖에 없는 상황이라 서로 모르도록 한 것이 맞는 방향인지 의문이 들었습니다.

### 3. 사용자 입력값을 TableView에 반영하기
- 사용자가 TextView에 입력한 내용을 TableView의 List에도 즉시 반영하기 위해 TextViewDelegate의 shouldChangeTextIn 메서드를 활용했습니다. 하지만 확인해보니 사용자 입력값이 1개 문자씩 delay되어 반영되는 문제가 발생하여 textViewDidChange 메서드를 활용하도록 변경했습니다. 

## 질문할 점

### 1. ActivityView의 타이틀이 업데이트되지 않는 문제
`ActivityView의 타이틀` 부분에 `메모의 타이틀`을 넣어주고 싶었습니다.

UIActivityViewController의 activityItems 파라미터에 title만을 넣어주면 ActivityView의 타이틀 부분에 나오긴 하지만, ActivityView의 액션(Copy 등)에서 사용할 아이템이 title밖에 없기에, Copy 버튼을 누를시 메모의 타이틀만 복사가 되는 문제가 있었습니다. (사진1)

Copy 버튼을 누를시 title과 title밑의 text까지 복사가 되길 원하여 아이템에 두가지 모두 넣어주었더니 Copy는 원하는대로 동작하지만, 액티비티 뷰의 타이틀 부분에 출력되는 텍스트들이 원하는대로 나오지 않았습니다. (사진2)

![Untitled](https://s3.us-west-2.amazonaws.com/secure.notion-static.com/dd069a6d-d9e5-44e3-9df8-80ba7bb0e020/Untitled.png?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Content-Sha256=UNSIGNED-PAYLOAD&X-Amz-Credential=AKIAT73L2G45EIPT3X45%2F20220216%2Fus-west-2%2Fs3%2Faws4_request&X-Amz-Date=20220216T124129Z&X-Amz-Expires=86400&X-Amz-Signature=8ca58df27e588405111548f63a75b0233222fb89149ffe2009356762f71b3b16&X-Amz-SignedHeaders=host&response-content-disposition=filename%20%3D%22Untitled.png%22&x-id=GetObject)

![Untitled](https://s3.us-west-2.amazonaws.com/secure.notion-static.com/ccd0bc8d-9362-4023-a556-8374f70de6d6/Untitled.png?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Content-Sha256=UNSIGNED-PAYLOAD&X-Amz-Credential=AKIAT73L2G45EIPT3X45%2F20220216%2Fus-west-2%2Fs3%2Faws4_request&X-Amz-Date=20220216T124114Z&X-Amz-Expires=86400&X-Amz-Signature=d5c4985c2453fce9b9ad5f9fe42580321cc3aeeecacbeb01223c6ed03a7db340&X-Amz-SignedHeaders=host&response-content-disposition=filename%20%3D%22Untitled.png%22&x-id=GetObject)
    
검색해보니 공유할 내용에 대한 meta 데이터(현재 저희 상황에선 title, text)를 수정하려면 [LPLinkMetadata](https://developer.apple.com/documentation/linkpresentation/lplinkmetadata)를 사용해야 하는데, 공식문서를 보면..

>An object that contains metadata about a URL.

라고 써있어서 현재 URL을 사용하지 않는 저희 상황에선 오버엔지니어링이라고 느껴져 사용하지 않았습니다.
    
```swift
// ActivityView 나타내기
let shareAction = UIAlertAction(title: "Share...", style: .default) { _ in
    let title = "Jokes Are Us Diagnostics:"
    let text = "Some Text"

    // 공유할 대상의 타입에 따라 Activity가 바뀜 (이미지 추가하면 연락처 공유도 뜸)
    let textToShare: [Any] = [ title, text ]  // 문제 : ActivityView의 타이틀에 임의 값이 들어가있음
    let textToShare2: [Any] = [ title ]  // 문제 : Bold 아닌 title로 뜸. Copy Activity 누르면 메모 내용 붙여넣기가 안됨

    let acitivityView = UIActivityViewController(activityItems: textToShare, applicationActivities: nil)

    let alertPopover = acitivityView.popoverPresentationController
    alertPopover?.barButtonItem = self.navigationItem.rightBarButtonItem

    self.present(acitivityView, animated: true, completion: nil)
}
```

현업에서는 액티비티뷰의 타이틀때문이라도 LPLinkMetaData를 사용하는지 궁금합니다.

아니시라면 혹시 이런 상황에선 어떻게 문제를 해결하시는지 궁금합니다!

    
### 2. iPhone 메모 앱의 ActivityView에서 썸네일 구현 방법

<img src="https://user-images.githubusercontent.com/55880539/154267566-223b7503-681b-46b6-9b85-bf38e22d6a76.png" alt="photo" height="400" width="220">

위의 사진과 같이 ActivityView를 보면 썸네일이미지, title, description이 들어간 것을 확인할 수 있습니다. 이때 썸네일이미지를 어떻게 구현해야 할지 의견을 나누어보았지만, 뚜렷한 결론을 내리지못해 질문 드립니다. (썸네일이미지의 배경색이 다크모드 일 때는 검정색으로, 라이트모드 일 때는 흰색으로 나타나는 것도 확인했습니다.)


### 3. 특정 상황에서 selectRow가 동작하지 않는 문제

`MasterTableViewController`에 위치한 `selectRow`가 담긴 `selectFirstCell` 메서드가 다른 곳에서는 정상작동하는데, swipe로 셀을 삭제할 때만 동일한 `selectFirstCell` 메서드가 호출됨에도 불구하고 첫 번째 Cell 선택 상태가 되지않고, scroll이 최상단으로 올라가지 않는 두 가지 문제가 있었습니다. 

분명 같은 메서드를 활용하여 첫번째 셀을 선택하게하고, 스크롤이 선택된 셀로 위치하도록 하였는데.. 이러한 문제가 왜 발생하는지를 찾아내지 못하여 질문드립니다.



### 4. textView에서 사용자가 수정한 내용을 CoreData에 반영하는 부분
- a. 현재 저희는 `shouldChangeTextIn` 메서드에서 `\n`으로 구분하여 TextView의 첫번째 줄은 `title`로, 그 이외에는 `body`로 설정해주어 CoreData에 저장을 해주고 있습니다. 헌데, 이러한 방법이 과연 효율적인 방법인지 다른 좋은 방법이 있을지 궁금합니다. 
- b. 사용자가 입력한 TextView에서 Title과 Body를 구분하고 효과를 주기위해서 textStorage라는 것을 사용했습니다. 처음에는 합쳐서 할당을 해주고 싶었는데 방법을 잘 못찾았습니다. 혹시 이렇게 하는 것 말고 다른 좋은 방법이 있을까 해서 질문드립니다.
        
```swift
func updateTextView() {
    let titleAttributes = [NSAttributedString.Key.font: UIFont.preferredFont(forTextStyle: .title1)]
    let bodyAttributes = [NSAttributedString.Key.font: UIFont.preferredFont(forTextStyle: .body)]

    let titleAttributedText = NSAttributedString(string: memo.title!, attributes: titleAttributes)
    let bodyAttributedText = NSAttributedString(string: "\n\(memo.body!)", attributes: bodyAttributes)

    textView.attributedText = titleAttributedText // title + body AttributedText를 합쳐서 할당해주고 싶었는데 안됨...
    textView.textStorage.append(bodyAttributedText)
}
        
extension DetailViewController: UITextViewDelegate {
    func textView(_ textView: UITextView, shouldChangeTextIn range: NSRange, replacementText text: String) -> Bool {
        let titleAttributes = [NSAttributedString.Key.font: UIFont.preferredFont(forTextStyle: .title1)]
        let bodyAttributes = [NSAttributedString.Key.font: UIFont.preferredFont(forTextStyle: .body)]

        let text = textView.text as NSString
        let titleRange = text.range(of: "\n")  // titleRange : 줄바꿈이 처음 나올 때까지의 range

        if titleRange.location >= range.location {  // range.location : 현재 입력한 텍스트 (range 매개변수)의 location 
            self.textView.typingAttributes = titleAttributes
        } else {
            self.textView.typingAttributes = bodyAttributes
        }

        return true
    }
}
```
### 5. ActivityView가 화면 중앙에 배치되지 않는 문제

셀을 스와이프하면 나오는 공유 버튼을 누르면 ActivityView가 띄워지는데, 해당 뷰를 화면 중앙에 배치하고 싶었습니다. 

acitivityView.popoverPresentationController의 sourceRect를 UIScreen을 활용하여 화면 중앙에 배치시키려는 등의 시도를 해봤지만.. 저희가 생각하는 위치와 다르게 배치되는 문제가 있었습니다.

[https://stackoverflow.com/questions/31759615/how-to-center-a-popoverview-in-swift](https://stackoverflow.com/questions/31759615/how-to-center-a-popoverview-in-swift)

그래서 현재는 cell의 옆에서 activityView가 띄워지도록 구성하였는데, 이를 화면 중앙에 띄워지게 하려면 좌표를 어떻게 줘야하는지 궁금합니다..
```swift
// 화면 중앙에 나타나지 않음
let alertPopover = acitivityView.popoverPresentationController
alertPopover?.sourceView = self.view
alertPopover?.sourceRect = CGRect(x: UIScreen.main.bounds.midX, y: UIScreen.main.bounds.midY, width: 0, height: 0) 
```


## 조언을 듣고싶은 점
### 1. 전체적인 네이밍과 가독성

전체적인 네이밍과 가독성이 적절한지 궁금합니다.

### 2. 코드 구조
구조에 대해 많이 고민했습니다. delegate 패턴, NotificationCenter 사용 등의 전체적인 코드 구조가 적절한지, 깔끔하게 구성이 된 것인지, 아니라면 어떤 점이 눈에 걸리시는지 정말 궁금합니다!!!

### 3. CoreData CRUD

이 프로젝트에선 크게 고려할 필요가 없었지만, 확장성을 높여주고 싶어 제네럴하게 CRUD 메서드들을 작성해보았습니다.

CoreData의 CRUD가 적절하게 작성된 것인지 궁금합니다. 

전체적으로 참고하면 좋을, 공부하면 좋을 것들도 전수해주시면 정말 감사하겠습니다!! 🙏

just1103 committed:
feat: detailButton 버튼에 대한 ActionSheet 및 ActivityView 구현 #3

- AlertFactory 타입 추가
- touchUpDetailButton 메서드 추가

yim2627 committed:
feat: CoreDataManger 및 CRUD 구현 #3

- CoreData Model 생성
- Json 파싱 절차를 CoreData 저장으로 변경
- 테이블 뷰 데이터소스가 가진 데이터 할당 방식 변경
- DetailView에 데이터 넘겨주는 방식 변경
- 버튼 액션함수 연결

oksunwoo committed:
feat: textView의 title과 body를 구분하여 효과 적용 #3

- CoreDataManager Generic 메서드로 개선
- 앱 실행시 첫번째 셀이 화면에 뜨도록 구현

just1103 committed:
feat: 사용자 메모 입력 시 CoreData 업데이트 구현 #3

- didChangeTextView Notification 추가를 통해 tableView reloadData 실행

yim2627 committed:
feat: tableView Swipe를 통한 공유, 삭제 기능 추가 #3

- 수정하고 있는 셀이 최상단으로 위치하도록 구현
- 수정날짜에 따라 재정렬되도록 CoreDataManager 수정
- tableViewDataSource가 가진 데이터 재정렬되도록 수정
- alert 구현

oksunwoo committed:
feat: 삭제 시 첫번째 셀 선택되도록 구현 #3

- DetailView에서 tableView로 삭제를 위한 Notification 구현
- UpdateMemo의 fetch 구현 변경

just1103 committed:
refactor: deleteTableViewCell 메서드 중복코드 제거 #3

yim2627 committed:
docs: STEP1 README 업데이트 #3

yim2627 committed:
docs: README 수정 #3

protocorn93 commented on CloudNotes/CloudNotes/Utility/CoreDataManager.swift:
```diff
@@ -0,0 +1,104 @@
+        } catch {
+            fatalError("\(error)")
+        }
+    }
```
다른 메서드와는 다르게 이 메서드는 제네릭하지 않네요. 그렇다면 다른 메서드들을 제네릭하게 작성한 이유가 없어져지 않을까요?!

just1103 replied:
더 범용적으로 사용할 수 있도록 제네릭 메서드를 적용해봤는데, 생각해보니 이번 프로젝트에서는 메모 데이터만 취급하기 때문에 굳이 제네릭을 할 필요가 없었다는 생각이 드네요.
(그리고 saveContext 메서드는 key 값들을 모두 매개변수로 처리해야하는 것, updateMemo/deleteMemo 메서드 등특정 memoId를 criteria로 저장된 데이터를 찾는 등의 데이터 관리도 프로젝트 특성에 따라 바뀌기 때문에 재사용성이 떨어진다고 생각했습니다.)

protocorn93 commented on CloudNotes/CloudNotes/Utility/CoreDataManager.swift:
```diff
@@ -0,0 +1,104 @@
+        } catch {
+            fatalError("\(error)")
+        }
+    }
```
이 메서드도 위 메서드와 같은 문제를 갖고있겠네요!

protocorn93 commented on CloudNotes/CloudNotes/Utility/CoreDataManager.swift:
```diff
@@ -0,0 +1,104 @@
+            
+            try context.save()
+        } catch {
+            fatalError("\(error)")
```
아직 에러 핸들링 코드는 작성되지 않은거죠!?

oksunwoo replied:
아직 구현하지 않았고, 다음 스탭에서 에러핸들링 추가로 구현해보겠습니다.👍

protocorn93 commented on CloudNotes/CloudNotes/Controller/MasterTableViewDataSource.swift:
```diff
@@ -1,25 +1,17 @@
 
 protocol MasterTableViewDataSourceProtocol: AnyObject {
-    var memos: [Memo]? { get }
+    var memos: [Memo]? { get set }
```
`memos`를 외부로 노출시킬 필요는 없어보입니다. `fetchData` 처럼, 삭제 혹은 아이템을 가져오는 인터페이스를 추가하는게 더 좋은 방법입니다!

just1103 replied:
`protocol` 및 `protocol을 채택한 타입`의 접근제어자 관계에 따라 protocol의 memos를 private으로 설정해도, 타입 내부의 memos를 최대 fileprivate으로 설정할 수 있었습니다. 그래서 타입 내부의 memos를 은닉화 (private으로 설정)하기 위해 protocol의 memos를 삭제했습니다. (다른 좋은 방법이 있을까요?)
그리고 ViewController에서 memos에 접근하는 내용은 메서드를 활용하도록 변경했습니다. 더 안전해진 것 같아요. 👍 

+) 아래 댓글로 남긴 CoreData CRUD의 제네릭 메서드 생성 관련 문제를 해결하면 MasterTableViewDataSourceProtocol의 구조 (CoreData 관련 메서드 및 memos 접근 메서드를 분리) 를 더 개선할 수 있을 것 같습니다.

protocorn93 commented on CloudNotes/CloudNotes/Controller/DetailViewController.swift:
```diff
@@ -1,6 +1,8 @@
 import UIKit
 
 final class DetailViewController: UIViewController {
+    var memo: Memo?
```
코드 구조상 이 프로퍼티는 `private`으로 정의되어도 무방해보이네요

just1103 replied:
PR 메시지로 말씀해주신 내용 중
> 먼저 리뷰에 남기지 못하 코멘트 하나 추가할게요! 이전 스탭 리뷰에서도 남긴 것 같은데 아직 반영이 되지 않은 것 같아서요.
MasterTableViewController의 memoDataSource 프로퍼티는 private let그리고 논옵셔널로 선언될 수 있어 보입니다. convenience init을 대신 init(..)을 사용해보시면 될 것 같아요.

이 부분을 개선하기 위해 아래 코드로 작성해봤는데, init 관련 에러가 발생해서 개선하지 못했습니다..
말씀하신 게 이러한 방향이 맞을까요?
![image](https://user-images.githubusercontent.com/70856586/154952284-d22149d5-f38f-4ae1-b081-82989270bbe5.png)

just1103 replied:
외부로 노출되지 않아서 private을 붙였습니다.

protocorn93 commented on CloudNotes/CloudNotes/Controller/DetailViewController.swift:
```diff
@@ -26,6 +29,10 @@ final class DetailViewController: UIViewController {
     
+    private func setupTextView() {
+        textView.delegate = self
+    }
```
이 코드는 `configureTextView`안에 들어가도 무방해보입니다.

protocorn93 commented on CloudNotes/CloudNotes/Controller/DetailViewController.swift:
```diff
@@ -59,14 +101,71 @@ final class DetailViewController: UIViewController {
+        var titleText: String = ""
+        var bodyText: String = ""
+        
+        if texts.count == 2 {
+            titleText = strings[0]
+            bodyText = strings[1]
+        } else if texts.count == 1 {
+            titleText = strings[0]
+        }
```
여기 작성된 `if-else-if` 구문을 보면, `count`를 먼저 검사하는 이유가 인덱스에 안전하게 접근하기 위함으로 보입니다. 이렇게 `count`를 검사할 수도 있지만 `Array`를 확장해서 subscript를 추가해주어 `strings[safe:0]`와 같이 사용할 수 있어요. 이를 사용하면 유효하지 않은 인덱스에 접근을 했을 때 옵셔널을 반환합니다. 

이를 활용해보는건 어떨까요?

yim2627 replied:
먼저 알려주셔서 감사합니다! 
하지만 해당 카운트를 검사하는 코드는 title과 body를 "\n"으로 나누는데, 타이틀만 있고, 바디는 없을 때를 대비하여 만든 조건문입니다.
그래서 알려주신 subscript를 사용하는 상황과 약간 다른 것 같다고 생각이 드는데, 콘 생각은 어떠신가요?

protocorn93 replied:
아하 그렇군요! 그런의도라면 get-only 프로퍼티 혹은 `fileprivate extension`을 활용해서 네이밍을 부여해줄 수 있을 것 같기도하네요! 의도는 이해 했습니다! 👍

protocorn93 commented on CloudNotes/CloudNotes/Controller/DetailViewController.swift:
```diff
@@ -59,14 +101,71 @@ final class DetailViewController: UIViewController {
+    func updateCoreData() {
+        let texts = textView.text.split(separator: "\n", maxSplits: 1)
+        let strings = texts.map { String($0) }
+        var titleText: String = ""
+        var bodyText: String = ""
```
이건 제가 궁금한 부분인데요. `titleText`와 `bodyText`가 옵셔널 타입을 가지면 안되는걸까요??

yim2627 replied:
텍스트가 없을 수 도 있는 상황이니 옵셔널 타입을 가져도 될 것 같습니다! 감사합니다 ㅎㅎ

protocorn93 commented on CloudNotes/CloudNotes/Model/TemporaryMemo.swift:
```diff
@@ -0,0 +1,8 @@
+    let body: String?
+    let lastModifiedDate: Double
+    let memoId: UUID
+}
```
`TemporaryMemo`의 용도를 설명해주실 수 있으실까요?

oksunwoo replied:
CoreData Class 인 Memo를 initialize 할 수가 없어서, Memo를 추가하거나 업데이트할 때 사용하기위해서 추가했습니다. 혹시 CoreData Class 인스턴스를 초기화할 수 있는 방법이 있을까요?

protocorn93 replied:
아~ 중간 매체 모델이라고 이해를 하면 되겠네요! 그렇다면 이 모델을 `Memo`로 가져가고 실제 코어데이터에 저장되는 타입을 다른 네이밍으로 가져가도 좋을 것 같아요 `MemoEntity` 라던지!

모델도 계층에 따라 구분해서 정의해놓으면 확장성 있는 구조에 도움이 됩니다.

protocorn93 commented on CloudNotes/CloudNotes/Controller/DetailViewController.swift:
```diff
@@ -46,10 +53,45 @@ final class DetailViewController: UIViewController {
+        let deleteAction = UIAlertAction(title: "Delete", style: .destructive) { _ in
+            let cancelAction = UIAlertAction(title: "취소", style: .cancel, handler: nil)
+            let deleteAction = UIAlertAction(title: "삭제", style: .destructive) { _ in
+                CoreDataManager.shared.deleteMemo(memoId: self.memo?.memoId)
```
`CoreDataManager.shared`를 사용하고 있는 모든 곳에 대한 공통 피드백입니다. 

싱글톤 객체를 사용하면 편합니다. 다만 사용하는 객체는 이 객체와 강하게 결합됩니다. 그렇기 때문에 테스트 코드 작성이 거의 불가능합니다. 그렇기 때문에 싱글톤 객체를 보고 "편리함을 위해 투명함을 희생한다"는 표현을 사용하곤 합니다. 

1. `CoreDataManager`가 정말 싱글톤으로 존재해야하는 이유에 대해 여러분들의 의견을 듣고 싶습니다. 
2. 또한 싱글톤 객체가 좀 더 유연하게 사용되기 위해선 이 역시 외부에서 주입해주는 방식으로 가져가는 것이 맞다고 생각합니다.

yim2627 replied:
CoreDataManager를 싱글톤으로 존재하게 한 이유는 말씀해주신 것처럼 여러 곳에서 간단하게 접근하여 저장하고, 수정할 수 있다는 편리함때문이였습니다.

하지만 static으로 인해 코어데이터를 사용하고 있지 않아도 메모리에 계속 상주하고 있다는 단점이 있으므로 DataSource에서만 코어데이터에 접근하여 사용하게하여 싱글톤을 제거하였습니다.

protocorn93 commented on CloudNotes/CloudNotes/Controller/MasterTableViewController.swift:
```diff
@@ -30,22 +30,88 @@ final class MasterTableViewController: UITableViewController {
+        
+        tableView.moveRow(at: previousIndexPath, to: firstIndexPath)
+        tableView.reloadRows(at: [firstIndexPath], with: .automatic)
+        tableView.selectRow(at: firstIndexPath, animated: true, scrollPosition: .none)
```
텍스트가 수정될 때마다 위의 코드가 실행되는 것으로 보이는데, 코드들의 목적이 이해가 가질 않네요. 

그리고 이렇게 되면 한글자 입력이 될 때마다 셀을 reload하고 있습니다. 그로인해 값이 입력될 때마다 셀이 번쩍(?)이게 되느네요.  이보단 셀의 직접 접근하여 값을 직접 갱신시켜주는게 좋지 않을까요?

just1103 replied:
reload를 삭제하고 Cell의 applyData 메서드를 활용하는 것으로 수정했습니다. 이제 깜빡이지 않네요!

protocorn93 commented on CloudNotes/CloudNotes/Controller/MasterTableViewController.swift:
```diff
@@ -30,22 +30,88 @@ final class MasterTableViewController: UITableViewController {
+    
+    @objc func touchUpAddMemoButton() {
+        let currentTime = NSDate().timeIntervalSince1970
+        let memo = TemporaryMemo(title: "새로운 메모", body: "테스트용 메모입니다.", lastModifiedDate: currentTime, memoId: UUID()) // Test
```
실제로 버튼을 누르고 새 메모가 추가되는 기능이기에, "테스트용 메모입니다."라는 문구는 제거되도 무방해보입니다.

just1103 replied:
다음 STEP에서 `메모추가 화면`을 구현할 줄 알고 테스트용 메모라고 입력해두었는데, STEP4까지 확인해보니 따로 요구사항으로 정의되어있지 않아서 콘 말씀대로 삭제해도 될 것 같습니다!

protocorn93 commented on CloudNotes/CloudNotes/Controller/MasterTableViewController.swift:
```diff
@@ -30,22 +30,88 @@ final class MasterTableViewController: UITableViewController {
+        let firstIndexPath = IndexPath(row: 0, section: 0)
+        
+        CoreDataManager.shared.saveContext(memo: memo)
+        memoDataSource?.memos = CoreDataManager.shared.fetch(request)
```
이 관계를 보았을 때, 

`MasterTableViewDataSource`는 `CoreDataManager`를 의존성으로 갖는게 좋아보입니다. `MasterTableViewDataSource`에는 적절한 이벤트만 전달하고 CRUD에 관한 로직은 이 객체 안에서 `CoreDataManager`를 통해 이루어지는게 좋겠죠. 

`MasterTableViewController`가 `CoreDataManager`에 대한 존재를 모르도록 하는게 좋아보이네요.

protocorn93 replied:
자 그렇다면 이제 여기서 `MasterTableViewDataSource`가 어떤 역할을 하는 것인지에 대해 고민을 하게 될 거에요.

처음에는 정말 "데이터소스"의 역할만 가져가는 것으로 생각하셨을거에요. 뭔가 제 피드백을 읽다보면 더 많은 역할을 가져가게 되는 것 같죠! 제가 지속해서 이렇게 피드백을 드리는 이유는 

1. 뷰컨트롤러의 역할을 최소화하기 위함입니다. 뷰컨트롤러는 뷰로써의 역할만을 담당하여 뷰의 레이아웃이나 사용자 이벤트 전달을 해주는 역할로 한정해야 합니다. 
2. 그렇다면 이러한 사용자 이벤트를 받아 각종 유틸 혹은 서비스 객체에게 다시 전달하고, 이들의 이벤트 처리 결과를 받아 다시 적절한 형태로 가공해서 뷰/뷰컨트롤러에게 돌려주는 역할을 하는 중간 매개체 역할을 하는 객체가 필요하겠죠. 

정답을 말씀드리자면 M-V-VM의 모습으로 객체를 나눠보시는 연습을 한다면 보다 깔끔하고 명확한 역할을 갖는 객체들로 앱을 만들어나갈 수 있습니다.

just1103 replied:
말씀해주신 것처럼 CoreDataManager를 `MasterTableViewDataSource`에 의존성 주입하고, CoreData CRUD 기능을 `MasterTableViewDataSource` 내부에 구현하도록 하면 ViewController에서 직접 CoreDataManager에 접근할 필요가 없어서 더욱 역할 분리가 명확해질 것 같습니다! 
이 과정에서 CoreData 클래스인 Model 타입과 TemporaryMemo 타입을 모두 사용하기 위해 `MemoEntity` 프로토콜을 추가하여 추상화해줬습니다.

protocorn93 commented on CloudNotes/CloudNotes/Utility/CoreDataManager.swift:
```diff
@@ -0,0 +1,104 @@
+import Foundation
+import CoreData
+
+class CoreDataManager {
```
`CoreDataManager`는 추상화가 가능해보입니다. 먼저 `CoreDataManager`의 역할에 대해 생각해보자구요!  `CoreDataManager`의 역할을 생각해보면 코어데이터라는 "외부저장소"에 데이터를 "관리"하는 역할을 합니다. 

추상화를 진행할 때는 변하는 것(변할 수 있는 것)과 변하지 않는 것을 구분할 필요가 있어요. 여기서 변할 수 있는 것은 무엇이며, 변하지 않는 것은 무엇일까요? 이것들을 생각해보는 것에서 추상화는 시작됩니다!

just1103 replied:
DM으로 알려주신 것처럼 `StorageProtocol`을 추가하여 CRUD 기능 (변하지 않는 것)을 정의하고,
`MemoCoreDataManager` 클래스 (변할 수 있는 것)가 해당 프로토콜을 채택하도록 하도록 했습니다.
(원래 `CoreDataManagerProtocol`도 추가하고, 제네럴 메서드로 CoreData CRUD 기능을 기본 구현하려고 했지만, Memo 데이터만 취급하고 있는 현재 프로젝트에서는 무의미할 것 같아서 이 과정은 생략했습니다.)

`MasterTableViewDataSource` 내부에 `memoDataManager` 프로퍼티를 추가하며, 프로퍼티 타입을 `StorageProtocol`로 지정 (이니셜라이저에서 `MemoCoreDataManager` 인스턴스를 기본값으로 할당)해주면 의존성을 낮출 수 있을 것 같아서 시도해봤습니다.

protocorn93 reviewed (changes requested):
쉽지 않은 프로젝트인데 세 분 모두 수고 많으셨어요!! 👏

먼저 리뷰에 남기지 못하 코멘트 하나 추가할게요! 이전 스탭 리뷰에서도 남긴 것 같은데 아직 반영이 되지 않은 것 같아서요. 

1. `MasterTableViewController`의 `memoDataSource` 프로퍼티는 `private let`그리고 논옵셔널로 선언될 수 있어 보입니다. `convenience init`을 대신 `init(..)`을 사용해보시면 될 것 같아요. 

---

## 구현내용 및 고민한 점 

### 1. 범용성 있는 CoreDataManager

이 부분은 코멘트를 확인해보시길 바랍니다. 

### 2. MasterTableViewController 및 DetailViewController의 관계

맞습니다. 이런 부분에 결국 답은 없습니다. 여러분들이 각각의 장단점을 고려해보시고 선택하셔야해요. 이러한 과정을 트레이드오프 과정이라고 합니다. 두 객체 관계를 고려해보았을 때, 이 둘의 결합도를 어디까지 고려해보아야할까라는 고민을 해보시는겁니다. 

말씀하신대로 기능과 프로젝트 특성상 두 객체의 연관성은 매우 높을 수 밖에 없습니다. 그렇기 때문에 둘은 서로의 존재를 알아야 합니다. 하지만 이말이 곧 두 객체 타입을 알아야 한다는 뜻은 아닙니다. 직접 접근이 아닌 간접적으로 접근할 수 있는 방법에 대해 고려해보시면 좋을 것 같아요. 

---

## 질문할 점 

### 1.  ActivityView의 타이틀이 업데이트되지 않는 문제

여기서 우리는 URL을 복사하는것이 아니기에 `LPLinkMetadata`를 사용할 필요는 없어보입니다. 단순히 텍스트를 복사하는 것이기에 현재 방식을 유지하는게 좋아보입니다. 

제가 담당해서 개발했던 기능에서도 동일하게 텍스트를 복사할 일이 있었지만, 여러분과 동일하게 만들어 사용했습니다!  

### 2. iPhone 메모 앱의 ActivityView에서 썸네일 구현 방법

이미지를 넣는 것은 저도 해보지는 않았지만, 이 [링크](https://stackoverflow.com/questions/57850483/ios13-share-sheet-how-to-set-preview-thumbnail-when-sharing-uiimage)에서 방법을 제시하고 있습니다.

하지만 예시를 들어준 이미지는 해당 이미지가 특별해서 다크모드/라이트모드를 지원하는 것이 아니라 단순히 화면을 캡처했고, 이를 사용하고 있는 것으로 보입니다. 

### 3. 특정 상황에서 selectRow가 동작하지 않는 문제

두 삭제 상황의 차이점이라하면 셀이 스와이프된 상태에서 삭제되려하기 때문이라고 생각하는데요. 결국 UI의 타이밍 문제이기에 결국 애니메이션이 끝난 후 콜백을 받아야할 것 같아요. 이를 위해선 CATransaction이나 의도적으로 delay를 주는 방법이 있을 것 같습니다. 

### 4. textView에서 사용자가 수정한 내용을 CoreData에 반영하는 부분

- 제가 생각했을 때도 구분을 위해선 최초 `\n`로 구분해주는 것이 최선인 것 같아보입니다. 
- `titleAttributedText`를 `NSMutableAttributedString`으로 정의해주면 `append` 메서드를 사용할 수 있습니다! 


### 5. ActivityView가 화면 중앙에 배치되지 않는 문제

기본 메모앱과 동일한 구조로 가져가길 희망하신다면 popOverView는 적절하지 않은 것 같아요. 반드시 화살표를 나타내야하는 것으로 보입니다. 이는 [HIG 문서](https://developer.apple.com/design/human-interface-guidelines/ios/views/popovers/)를 통해서 확인해볼 수 있습니다. 

이와 같은 고민은 이 [트위터 스레드](https://twitter.com/steipete/status/755409285111218180?lang=bg)에서도 확인할 수 있습니다. 그렇기에 전통적인 popOverView로는 적절하지 않은 것 같고 alertController와 함께 커스터마이징을 해야 가능하지 않을까 하는 생각을 해봅니다. 

현재 단계에선 셀에 popOverView를 두는 것이 최선으로 보입니다!

oksunwoo committed:
feat: CoreDataManager 싱글톤 삭제 #4

- TableViewDatasource에 CRUD 추가
- Notification으로 userInfo 전달

just1103 committed:
refactor: StorageProtocol 추가하여 Memo 및 TemporaryMemo 타입 추상화 #4

- updateTableView 메서드 개선

yim2627 committed:
fix: 메모 업데이트시 캐스팅 타입이 달라 함수를 탈출하는 버그 수정 #4

- warning 제거
- Memo가 비어있는지 화인해 DetailView를 업데이트하는  changeMemoIfNotEmpty 메서드 구현

oksunwoo committed:
refactor: DataSource memos 은닉화 #4

just1103 committed:
feat: DropBox 연동 및 업로드/다운로드 메서드 추가 #5

- plist 추가
- DropBox 토큰 및 client 인스턴스 생성
- DropBox에 폴더 생성, 데이터 업로드/다운로드 메서드 추가

just1103 committed:
chore: DropBox 관련 메서드 주석 처리 #5

yim2627 committed:
refactor: MemoCoreDataManager, DataSource 코드 구조 수정 #4

- MemosManageable, CoreDataManageable 추가하여 역할 분리
- MasterTableViewController 이니셜라이저 수정
- 전반적인 코드 정리
